docs/data/data-grid/row-updates/LazyLoadingGrid.tsx

```diff
@@ -0,0 +1,51 @@
+import * as React from 'react';
+import {
+  DataGridPro,
+  GridFetchRowsParams,
+  useGridApiRef,
+} from '@mui/x-data-grid-pro';
+import { createFakeServer, UseDemoDataOptions } from '@mui/x-data-grid-generator';
+
+const ROW_COUNT = 10000;
+const DATASET_OPTION: UseDemoDataOptions = {
+  dataSet: 'Employee',
+  rowLength: ROW_COUNT,
+};
+
+const { columns, useQuery } = createFakeServer(DATASET_OPTION);
+
+export default function LazyLoadingGrid() {
```

Comment:
This demo has a couple of bugs:

1. Filtering is not supported
2. Sorting gets outdated

I have looked into a fix, it can be done with:

```diff
diff --git a/docs/data/data-grid/row-updates/LazyLoadingGrid.tsx b/docs/data/data-grid/row-updates/LazyLoadingGrid.tsx
index 79dbe9813..fa22ebcf6 100644
--- a/docs/data/data-grid/row-updates/LazyLoadingGrid.tsx
+++ b/docs/data/data-grid/row-updates/LazyLoadingGrid.tsx
@@ -4,30 +4,65 @@ import {
   GridFetchRowsParams,
   useGridApiRef,
 } from '@mui/x-data-grid-pro';
-import { createFakeServer, UseDemoDataOptions } from '@mui/x-data-grid-generator';
+import {
+  createFakeServer,
+  QueryOptions,
+  loadServerRows,
+  UseDemoDataOptions,
+} from '@mui/x-data-grid-generator';

-const ROW_COUNT = 10000;
 const DATASET_OPTION: UseDemoDataOptions = {
   dataSet: 'Employee',
-  rowLength: ROW_COUNT,
+  rowLength: 10000,
 };

-const { columns, useQuery } = createFakeServer(DATASET_OPTION);
+const { columns, columnsWithDefaultColDef, useQuery } =
+  createFakeServer(DATASET_OPTION);

 export default function LazyLoadingGrid() {
   const apiRef = useGridApiRef();
-  const cachedQuery = React.useMemo(() => ({}), []);
-  const { data, loadServerRowsInterval } = useQuery(cachedQuery);
-  const cachedRows = React.useMemo(() => data.slice(0, 10), [data]);
+  const [queryOptions, setQueryOptions] = React.useState<QueryOptions>({});
+  const { data } = useQuery(queryOptions);
+
+  const initialRows = React.useMemo(() => data.slice(0, 10), [data]);
   const cachedColumns = React.useMemo(() => columns, []);

   const handleFetchRows = async (params: GridFetchRowsParams) => {
-    const newRowsBatch = await loadServerRowsInterval(params);
+    // Bug, it shouldn't be triggered for the initial rendering
+    if (data.length === 0) {
+      return;
+    }
+
+    setQueryOptions((prev) => {
+      if (
+        prev.filterModel === params.filterModel &&
+        prev.sortModel === params.sortModel
+      ) {
+        return prev;
+      }
+
+      return {
+        filterModel: params.filterModel,
+        sortModel: params.sortModel,
+      };
+    });
+
+    const serverRows = await loadServerRows(
+      data,
+      {
+        filterModel: params.filterModel,
+        sortModel: params.sortModel,
+      },
+      {
+        useCursorPagination: false,
+      },
+      columnsWithDefaultColDef,
+    );

     apiRef.current.replaceRows(
       params.firstRowToRender,
       params.lastRowToRender,
-      newRowsBatch.returnedRows,
+      serverRows.returnedRows.slice(params.firstRowToRender, params.lastRowToRender),
     );
   };

@@ -35,10 +70,10 @@ export default function LazyLoadingGrid() {
     <div style={{ height: 400, width: '100%' }}>
       <DataGridPro
         columns={cachedColumns}
-        rows={cachedRows}
+        rows={initialRows}
         apiRef={apiRef}
         hideFooterPagination
-        rowCount={ROW_COUNT}
+        rowCount={data.length}
         sortingMode="server"
         filterMode="server"
         rowsLoadingMode="server"
diff --git a/packages/grid/x-data-grid-generator/src/hooks/useQuery.ts b/packages/grid/x-data-grid-generator/src/hooks/useQuery.ts
index 84648e349..87478a253 100644
--- a/packages/grid/x-data-grid-generator/src/hooks/useQuery.ts
+++ b/packages/grid/x-data-grid-generator/src/hooks/useQuery.ts
@@ -107,7 +107,7 @@ const getFilteredRows = (
 /**
  * Simulates server data loading
  */
-const loadServerRows = (
+export const loadServerRows = (
   rows: GridRowModel[],
   queryOptions: QueryOptions,
   serverOptions: ServerOptions,
@@ -170,12 +170,14 @@ interface PageInfo {
   pageSize?: number;
 }

-export interface ServerOptions {
+interface DefaultServerOptions {
   minDelay: number;
   maxDelay: number;
   useCursorPagination?: boolean;
 }

+type ServerOptions = Partial<DefaultServerOptions>;
+
 export interface QueryOptions {
   cursor?: GridRowId;
   page?: number;
@@ -196,7 +198,7 @@ const DEFAULT_DATASET_OPTIONS: UseDemoDataOptions = {
 declare const DISABLE_CHANCE_RANDOM: any;
 const disableDelay = typeof DISABLE_CHANCE_RANDOM !== 'undefined' && DISABLE_CHANCE_RANDOM;

-const DEFAULT_SERVER_OPTIONS: ServerOptions = {
+const DEFAULT_SERVER_OPTIONS: DefaultServerOptions = {
   minDelay: disableDelay ? 0 : 100,
   maxDelay: disableDelay ? 0 : 300,
   useCursorPagination: true,
@@ -204,7 +206,7 @@ const DEFAULT_SERVER_OPTIONS: ServerOptions = {

 export const createFakeServer = (
   dataSetOptions?: Partial<UseDemoDataOptions>,
-  serverOptions?: Partial<ServerOptions>,
+  serverOptions?: ServerOptions,
 ) => {
   const dataSetOptionsWithDefault = { ...DEFAULT_DATASET_OPTIONS, ...dataSetOptions };
   const serverOptionsWithDefault = { ...DEFAULT_SERVER_OPTIONS, ...serverOptions };
@@ -275,43 +277,11 @@ export const createFakeServer = (
     // We use queryOptions pointer to be sure that isLoading===true as soon as the options change
     const effectShouldStart = queryOptionsRef.current !== queryOptions;

-    const loadServerRowsInterval = (
-      query: Pick<
-        QueryOptions,
-        'filterModel' | 'sortModel' | 'firstRowToRender' | 'lastRowToRender'
-      >,
-      server = serverOptionsWithDefault,
-    ): Promise<FakeServerResponse> => {
-      const { minDelay = 100, maxDelay = 300 } = server;
-
-      if (maxDelay < minDelay) {
-        throw new Error('serverOptions.minDelay is larger than serverOptions.maxDelay ');
-      }
-
-      const delay = Math.random() * (maxDelay - minDelay) + minDelay;
-      const { firstRowToRender, lastRowToRender } = query;
-      let filteredRows = getFilteredRows(rows, query.filterModel, columnsWithDefaultColDef);
-      const rowComparator = getRowComparator(query.sortModel, columnsWithDefaultColDef);
-      filteredRows = [...filteredRows].sort(rowComparator);
-      const totalRowCount = filteredRows.length;
-      const res: FakeServerResponse = {
-        returnedRows: filteredRows.slice(firstRowToRender, lastRowToRender),
-        totalRowCount,
-      };
-
-      return new Promise<FakeServerResponse>((resolve) => {
-        setTimeout(() => {
-          resolve(res);
-        }, delay); // simulate network latency
-      });
-    };
-
     return {
       isLoading: isLoading || effectShouldStart,
-      loadServerRowsInterval,
       ...response,
     };
   };

-  return { columns, initialState, useQuery };
+  return { columns, columnsWithDefaultColDef, initialState, useQuery };
 };
 ```

https://user-images.githubusercontent.com/3165635/186415286-ee025157-9c95-4d7b-8783-0a9a191a3bab.mov

Now, there is still something a bit off. When the network request is too fast, the Skeleton creates a blink.

Reply:
Ok thanks.